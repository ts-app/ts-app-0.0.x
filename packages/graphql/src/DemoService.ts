import { Resolver } from './Resolver'
import { ServiceInfo } from './ServiceInfo'

export class DemoService {
  private anotherNumber: number = 123

  constructor (public count: number = 0) {
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}"`
  }

  @Resolver({ type: 'mutation' })
  add ({ howMany = 1 }: { howMany?: number }) {
    this.count += howMany
    return this.count
  }

  @Resolver({ type: 'mutation' })
  async promiseToSubtract ({ howMany = 1 }: { howMany?: number }): Promise<number> {
    return new Promise<number>(resolve => {
      setTimeout(() => {
        this.count -= howMany
        resolve(this.count)
      }, 1)
    })
  }

  justUseAnotherNumber () {
    return this.anotherNumber * 3
  }

  @Resolver()
  throwSomeError () {
    throw new Error('simulated error')
  }
}

export class DemoServiceWithServiceInfo implements ServiceInfo {
  info (): { serviceName: string } {
    return {
      serviceName: 'CustomServiceName'
    }
  }

  @Resolver({ paramNames: [ 'name', 'message' ] })
  echo2 (name: string, message: string) {
    return `Hello, ${name}! Message is "${message}" - from echo2`
  }
}
