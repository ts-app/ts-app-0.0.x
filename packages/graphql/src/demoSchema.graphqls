type Query {
  echo (name: String!, message: String!): String
  throwSomeError: String
}

type Mutation {
  add (howMany: Int): Int
  promiseToSubtract (howMany: Int): Int
}
