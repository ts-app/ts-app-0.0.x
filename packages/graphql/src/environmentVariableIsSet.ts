import { ResolverMiddlewareInput } from './ResolverMiddlewareInput'

/**
 * This function is used to enable/disable access to GraphQL resolver based on specified environment
 * variable to be set to '1'.
 *
 * @param {string} name Environment variable name
 * @return {(input: ResolverMiddlewareInput) => boolean}
 */
export const environmentVariableIsSet = (name: string) => {
  const authorized = process.env[ name ] === '1'

  return (input: ResolverMiddlewareInput): boolean => authorized
// tslint:disable-next-line
}
