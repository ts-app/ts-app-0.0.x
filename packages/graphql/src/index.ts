export * from './demoSchemaDefinition'
export * from './DemoService'
export * from './environmentVariableIsSet'
export * from './mergeSchemaDefinitions'
export * from './mergeTypeDefs'
export * from './Resolver'
export * from './ResolverError'
export * from './ResolverMiddlewareInput'
export * from './ResolverService'
export * from './SchemaDefinition'
export * from './SecurityMiddleware'
export * from './serveSchema'
export * from './ServiceInfo'
export * from './standardSchemaDefinition'
export * from './userIdMatchParam'
