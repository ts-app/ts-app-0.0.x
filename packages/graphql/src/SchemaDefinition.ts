import { GraphQLFieldResolver, GraphQLScalarType } from 'graphql'

export type IResolverObject = {
  [key: string]: GraphQLFieldResolver<any, any>
}

export interface Resolvers {
  [key: string]: (() => any) | IResolverObject | GraphQLScalarType
}

export type SchemaDefinition = {
  typeDefs?: string,
  resolvers?: Resolvers,
  dependencies?: {
    [dependencyName: string]: {
      typeDefs?: string,
      resolvers?: Resolvers
    }
  }
}
