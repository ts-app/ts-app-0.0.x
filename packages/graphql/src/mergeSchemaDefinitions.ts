import { SchemaDefinition } from './SchemaDefinition'
import { mergeTypeDefs } from './mergeTypeDefs'

/**
 *
 * Merge an array of SchemaDefinition and return IExecutableSchemaDefinition (from Apollo) that can
 * be used as an input for makeExecutableSchema().
 *
 * It extracts contents from schema definition's "dependencies" to be returned as part of the
 * IExecutableSchemaDefinition. This allows reusing SchemaDefinition when specified as
 * "dependencies".
 *
 * @param {SchemaDefinition[]} definitions
 * @return {{typeDefs: string; resolvers: {Query: {}; Mutation: {}}}}
 */
export const mergeSchemaDefinitions = (definitions: SchemaDefinition[]) => {
  let typeDefs = ''
  let mergedResolvers = {
    Query: {},
    Mutation: {}
  }

  // flatten dependencies for all definitions
  const flatDependencies: SchemaDefinition[] = definitions.map(d => {
    if (d.dependencies) {
      return Object.keys(d.dependencies).map(name => d.dependencies![ name ])
    } else {
      return []
    }
  }).reduce((prev, current) => prev.concat(current), [])

  // resolve all definitions as "dependencies" plus provided "definitions"
  const allDefinitions = [ ...flatDependencies, ...definitions ]

  // perform merge
  allDefinitions.map(d => {
    // --- (simple) merge for typeDefs
    if (d.typeDefs) {
      if (typeDefs.length === 0) {
        typeDefs = d.typeDefs
      } else {
        typeDefs = mergeTypeDefs([ typeDefs, d.typeDefs ])
      }
    }

    if (d.resolvers) {
      mergedResolvers = {
        ...mergedResolvers,
        ...d.resolvers
      }

      if (d.resolvers.Query) {
        mergedResolvers = {
          ...mergedResolvers,
          Query: {
            ...mergedResolvers.Query,
            ...d.resolvers.Query
          }
        }
      }

      if (d.resolvers.Mutation) {
        mergedResolvers = {
          ...mergedResolvers,
          Mutation: {
            ...mergedResolvers.Mutation,
            ...d.resolvers.Mutation
          }
        }
      }
    }
  })

  return {
    typeDefs,
    resolvers: mergedResolvers
  }
  // tslint:disable-next-line
}
