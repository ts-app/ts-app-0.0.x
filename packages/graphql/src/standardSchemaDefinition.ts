import { GraphQLScalarType, Kind } from 'graphql'
import { SchemaDefinition } from './SchemaDefinition'

export const standardSchemaDefinition = (): SchemaDefinition => {
  // http://dev.apollodata.com/tools/graphql-tools/scalars.html
  const resolvers = {
    Date: new GraphQLScalarType({
      name: 'Date',
      description: 'Date custom scalar type',
      parseValue (value) {
        return new Date(value) // value from the client
      },
      serialize (value) {
        return value.getTime() // value sent to the client
      },
      parseLiteral (ast) {
        if (ast.kind === Kind.INT) {
          return parseInt(ast.value, 10) // ast value is always in string format
        }
        return null
      }
    })
  }

  const typeDefs = 'scalar Date'

  return {
    resolvers,
    typeDefs
  }
}
