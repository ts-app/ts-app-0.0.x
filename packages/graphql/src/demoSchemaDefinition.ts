import { loadFile } from '@ts-app/common'
import { DemoService } from './DemoService'
import { ResolverService } from './ResolverService'
import { standardSchemaDefinition } from './standardSchemaDefinition'
import { SchemaDefinition } from './SchemaDefinition'

export const demoSchemaDefinition = (): SchemaDefinition => {
  const resolver = ResolverService.getInstance()
  resolver.registerService(new DemoService())
  const typeDefs = loadFile(`${__dirname}/demoSchema.graphqls`)
  const resolvers = resolver.makeResolvers()
  const standard = standardSchemaDefinition()

  return {
    typeDefs,
    resolvers,
    dependencies: {
      standard
    }
  }
}
