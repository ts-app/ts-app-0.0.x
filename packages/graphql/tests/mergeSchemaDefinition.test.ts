import { mergeSchemaDefinitions } from '../src/mergeSchemaDefinitions'
import { SchemaDefinition } from '../src/SchemaDefinition'
import { standardSchemaDefinition } from '../src/standardSchemaDefinition'

describe('mergeSchemaDefinition', () => {
  test('simple merge', () => {
    const sd1: SchemaDefinition = {
      typeDefs: `
type Query {
  query1: String
}

type Mutation {
  mutation1: String
}
`,
      resolvers: {
        Query: {
          query1: () => 'query value'
        },
        Mutation: {
          mutation1: () => 'mutation value'
        }
      },
      dependencies: {
        standard: standardSchemaDefinition()
      }
    }
    const sd2: SchemaDefinition = {
      typeDefs: `
type Query {
  sd2Query1: Int
}

type Mutation {
  sd2Mutation1: Boolean
}
`,
      resolvers: {
        Query: {
          sd2Query1: () => 123
        },
        Mutation: {
          sd2Mutation1: () => true
        }
      }
    }

    const merged = mergeSchemaDefinitions([ sd1, sd2 ])
    expect(merged).toMatchSnapshot()
  })
})
