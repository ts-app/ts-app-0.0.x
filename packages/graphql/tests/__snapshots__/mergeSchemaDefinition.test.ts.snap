// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeSchemaDefinition simple merge 1`] = `
Object {
  "resolvers": Object {
    "Date": "Date",
    "Mutation": Object {
      "sd2Mutation1": [Function],
    },
    "Query": Object {
      "sd2Query1": [Function],
    },
  },
  "typeDefs": "type Mutation {

mutation1: String
sd2Mutation1: Boolean}
type Query {

query1: String
sd2Query1: Int}
scalar Date









",
}
`;
